sys/arm64/arm64/cheri_revoke_machdep.c

```diff
@@ -0,0 +1,414 @@
+		 * set or something.
+		 */
+
+		CHERI_REVOKE_STATS_BUMP(crst, caps_found_revoked);
+	} else if (cheri_gettag(cut) && ctp(crshadow, cut, perms, start, end)) {
+		void * __capability cscratch;
+		int stxr_status;
+
+		uintcap_t cutr = cheri_revoke_cap(cut);
+
+		CHERI_REVOKE_STATS_BUMP(crst, caps_found);
+
+		/*
+		 * Load-link the position under test; verify that it matches
+		 * our previous load; store conditionally the revoked
+		 * version back.  If the verification fails, don't try
+		 * anything fancy, just modify the return value to flag the
+		 * page as dirty.
+		 *
+		 * It's possible that this CAS will fail because the pointer
+		 * has changed during our test.  That's fine, if this is not
+		 * a stop-the-world scan; we'll catch it in the next go
+		 * around.  However, because CAS can fail for reasons other
+		 * than an actual data failure, return an indicator that the
+		 * page should not be considered clean.
+		 *
+		 * Because revoked capabilities are still tagged, one might
+		 * worry that this would reset the capdirty bits.  That's
+		 * not true, tho', because we're storing via the direct
+		 * mapping of physical memory.
+		 */
+again:
+		/*
+		 * stxr returns 0 or 1, so use a value of 2
+		 * to indicate that it was not executed.
+		 */
+		stxr_status = 2;
+
+		__asm__ __volatile__ (
+#ifndef __CHERI_PURE_CAPABILITY__
+			"bx #4\n\t"
+			".arch_extension c64\n\t"
+#endif
+			"ldxr %[cscratch], [%[cutp]]\n\t"
+			"cmp %[cscratch], %[cut]\n\t"
+			"bne 1f\n\t"
+			"stxr %w[stxr_status], %[cutr], [%[cutp]]\n\t"
+			"1:\n\t"
+#ifndef __CHERI_PURE_CAPABILITY__
+			"bx #4\n\t"
+			".arch_extension noc64\n\t"
+			".arch_extension a64c\n\t"
+#endif
+		  : [stxr_status] "=r" (stxr_status),
+		    [cscratch] "=&C" (cscratch), [cutr] "+C" (cutr)
+		  : [cut] "C" (cut), [cutp] "C" (cutp)
+		  : "memory");
+
+		/* stxr returns 0 on success */
+		if (__builtin_expect(stxr_status == 0, 1)) {
+			CHERI_REVOKE_STATS_BUMP(crst, caps_cleared);
+			/* Don't count a revoked cap as HASCAPS */
+		} else if (!cheri_gettag(cscratch)) {
+			/* Data; don't sweat it */
+		} else if (cheri_revoke_is_revoked(cscratch)) {
+			/* Revoked cap; don't worry about it */
+		} else if (__builtin_expect(stxr_status == 1, 1)) {
+			/* stxr returns 1 on failure */
+			goto again;
+		} else {
+			/*
+			 * An unexpected capability - stxr_status was neither 0
+			 * nor 1, which means that the stxr wasn't executed and
+			 * so the capability at cutp has changed.
+			 */
+			*res |= VM_CHERI_REVOKE_PAGE_DIRTY
+				| VM_CHERI_REVOKE_PAGE_HASCAPS ;
+		}
+	} else {
+		CHERI_REVOKE_STATS_BUMP(crst, caps_found);
+
+		/*
+		 * Even though it might actually be un-tagged, that's a very
+		 * narrow race and this a very common case, so don't bother
+		 * testing.  We'll find it clear next time, maybe.
+		 */
+		*res |= VM_CHERI_REVOKE_PAGE_HASCAPS;
+	}
+
+	return (0);
+}
+
+static inline void
+enable_user_memory_access()
```

Comment:
Sadly, these probably live "forever" on Morello, but it'd be good not to need them next time around.